File: .zed-plugins/macos/install-plugin.js
const tasks = ".zed-plugins/jsons-data/tasks.json";
const keymap = ".zed-plugins/jsons-data/keymap.json";

async function readFile(filePath) {
  if (typeof Deno !== "undefined") {
    return await Deno.readTextFile(filePath);
  } else if (typeof require !== "undefined") {
    const fs = (await import("fs")).promises;
    return await fs.readFile(filePath, "utf8");
  }
  throw new Error("Unsupported runtime");
}

async function writeFile(filePath, content) {
  if (typeof Deno !== "undefined") {
    await Deno.writeTextFile(filePath, content);
  } else if (typeof require !== "undefined") {
    const fs = (await import("fs")).promises;
    await fs.writeFile(filePath, content, "utf8");
  } else {
    throw new Error("Unsupported runtime");
  }
}

function updateTasks(tasks, newTasks) {
  const tasksMap = new Map(tasks.map((task) => [task.label, task]));
  newTasks.forEach((task) => {
    tasksMap.has(task.label)
      ? Object.assign(tasksMap.get(task.label), task)
      : tasksMap.set(task.label, task);
  });
  return Array.from(tasksMap.values());
}

function updateKeymaps(keymaps, newKeymaps) {
  const keymapMap = new Map(keymaps.map((k) => [k.context, k]));
  newKeymaps.forEach((newKm) => {
    const existing = keymapMap.get(newKm.context);
    if (existing) {
      existing.bindings = { ...existing.bindings, ...newKm.bindings };
    } else {
      keymapMap.set(newKm.context, newKm);
    }
  });
  return Array.from(keymapMap.values());
}

function removeJsonComments(jsonString) {
  return jsonString.replace(/\/\/.*$/gm, "").replace(/\/\*[\s\S]*?\*\//g, "");
}

async function installTasks(homeDir) {
  const tasksJsonPath = `${homeDir}/.config/zed/tasks.json`;
  const newTasksPath = tasks;

  try {
    const [existing, newTasks] = await Promise.all([
      readFile(tasksJsonPath).then(removeJsonComments).then(JSON.parse),
      readFile(newTasksPath).then(JSON.parse),
    ]);
    await writeFile(
      tasksJsonPath,
      JSON.stringify(updateTasks(existing, newTasks), null, 2),
    );
    console.log("✓ Tasks installed successfully");
  } catch (error) {
    console.error("Error installing tasks:", error);
  }
}

async function installKeymaps(homeDir) {
  const keymapJsonPath = `${homeDir}/.config/zed/keymap.json`;
  const newKeymapsPath = keymap;

  try {
    const [existing, newKeymaps] = await Promise.all([
      readFile(keymapJsonPath).then(removeJsonComments).then(JSON.parse),
      readFile(newKeymapsPath).then(JSON.parse),
    ]);
    await writeFile(
      keymapJsonPath,
      JSON.stringify(updateKeymaps(existing, newKeymaps), null, 2),
    );
    console.log("✓ Keymaps installed successfully");
  } catch (error) {
    console.error("Error installing keymaps:", error);
  }
}

function detectOS() {
  // 1. Deno detection
  if (typeof Deno !== "undefined") {
    return Deno.build.os === "darwin" ? "macos" : "linux";
  }

  // 2. Node.js detection
  if (typeof process !== "undefined") {
    const { platform } = require("os");
    return platform() === "darwin" ? "macos" : "linux";
  }

  // 3. Browser detection
  if (typeof navigator !== "undefined") {
    const platform = navigator.platform.toLowerCase();
    if (platform.includes("mac") || platform.includes("iphone")) return "macos";
    if (platform.includes("linux")) return "linux";
  }

  return "unknown";
}

(async () => {
  console.log("-> Installing OrganizeGPT plugin JS...");
  const homeDir =
    typeof Deno !== "undefined"
      ? Deno.env.get("HOME") || Deno.env.get("USERPROFILE")
      : (await import("os")).homedir();

  await Promise.all([installTasks(homeDir), installKeymaps(homeDir)]);
})();

File: .zed-plugins/macos/install-plugin.rb
require 'json'
require 'fileutils'

TASKS_PATH = ".zed-plugins/jsons-data/tasks.json"
KEYMAP_PATH = ".zed-plugins/jsons-data/keymap.json"

def read_file(file_path)
  File.read(file_path)
end

def write_file(file_path, content)
  FileUtils.mkdir_p(File.dirname(file_path))
  File.write(file_path, content)
end

def update_tasks(existing_tasks, new_tasks)
  tasks_map = existing_tasks.each_with_object({}) { |task, h| h[task['label']] = task }
  new_tasks.each do |task|
    if tasks_map.key?(task['label'])
      tasks_map[task['label']].merge!(task)
    else
      tasks_map[task['label']] = task
    end
  end
  tasks_map.values
end

def update_keymaps(existing_keymaps, new_keymaps)
  keymap_map = existing_keymaps.each_with_object({}) { |km, h| h[km['context']] = km }
  new_keymaps.each do |new_km|
    context = new_km['context']
    if keymap_map.key?(context)
      keymap_map[context]['bindings'] = keymap_map[context]['bindings'].merge(new_km['bindings'])
    else
      keymap_map[context] = new_km
    end
  end
  keymap_map.values
end

def remove_json_comments(json_string)
  json_string.gsub(/\/\/.*$/, '')
             .gsub(/\/\*.*?\*\//m, '')
end

def install_tasks(home_dir)
  tasks_json_path = File.join(home_dir, '.config/zed/tasks.json')
  new_tasks_path = TASKS_PATH

  begin
    existing = JSON.parse(remove_json_comments(File.read(tasks_json_path)))
    new_tasks = JSON.parse(File.read(new_tasks_path))
    merged = update_tasks(existing, new_tasks)
    write_file(tasks_json_path, JSON.pretty_generate(merged))
    puts "✓ Tasks installed successfully"
  rescue => e
    puts "Error installing tasks: #{e}"
  end
end

def install_keymaps(home_dir)
  keymap_json_path = File.join(home_dir, '.config/zed/keymap.json')
  new_keymaps_path = KEYMAP_PATH

  begin
    existing = JSON.parse(remove_json_comments(File.read(keymap_json_path)))
    new_keymaps = JSON.parse(File.read(new_keymaps_path))
    merged = update_keymaps(existing, new_keymaps)
    write_file(keymap_json_path, JSON.pretty_generate(merged))
    puts "✓ Keymaps installed successfully"
  rescue => e
    puts "Error installing keymaps: #{e}"
  end
end

def detect_os
  case RUBY_PLATFORM.downcase
  when /darwin/ then 'macos'
  when /linux/ then 'linux'
  else 'unknown'
  end
end

puts "-> Installing OrganizeGPT plugin with Ruby..."
puts ""
home_dir = Dir.home

install_tasks(home_dir)
install_keymaps(home_dir)
